src/backend/distributed/commands/multi_copy.c

```diff
@@ -0,0 +1,1284 @@
+/*-------------------------------------------------------------------------
+ *
+ * multi_copy.c
+ *     This file contains implementation of COPY utility for distributed
+ *     tables.
+ *
+ * The CitusCopyFrom function should be called from the utility hook to
+ * process COPY ... FROM commands on distributed tables. CitusCopyFrom
+ * parses the input from stdin, a program executed on the master, or a file
+ * on the master, and decides into which shard to put the data. It opens a
+ * new connection for every shard placement and uses the PQputCopyData
+ * function to copy the data. Because of buffering in PQputCopyData, the
+ * workers will ingest the data at least partially in parallel.
```

Comment:
Hm. Isn't that more due to the fact that PQputCopyData just hands of the data to the kernel and then continues? And network pipelining effects take care of "parallelizing"?